Figure out Tests for TransferFormWizard

yenaing-oo opened the pull request:
Closes #412

yenaing-oo commented:
I managed to set up a test class for testing the wizard, it turns out I didn't have to do anything other than login when making requests to it. I was under the impression that I had to mock the storage session used by the wizard.

Are there any other test cases you'd like me to add?

danloveg commented:
> I managed to set up a test class for testing the wizard, it turns out I didn't have to do anything other than login when making requests to it. I was under the impression that I had to mock the storage session used by the wizard.
> 
> Are there any other test cases you'd like me to add?

Nice! No need to add any more tests in this PR, I mostly just wanted to figure out how to test the form wizard up to the last step, and looks like you've got it. I'll submit my review later this afternoon or tomorrow morning.

yenaing-oo commented on app/recordtransfer/tests/unit/views/test_transfer.py:
```diff
@@ -15,32 +17,206 @@ def setUp(self) -> None:
-        self.assertEqual(response.status_code, 302)
-        self.assertEqual(response.url, reverse("recordtransfer:in_progress_submission_expired"))
+            if step != TransferStep.REVIEW.value and "form" in response.context:
+                self.assertFalse(response.context["form"].errors)
```
That's a good idea, I've added that here: https://github.com/NationalCentreTruthReconciliation/Secure-Record-Transfer/pull/442/commits/29ac7d279068b78c98aede87fadf35b55bd7281f

yenaing-oo commented on app/recordtransfer/tests/unit/views/test_transfer.py:
```diff
@@ -15,32 +17,206 @@ def setUp(self) -> None:
+                self.assertFalse(response.context["form"].errors)
+
+    @patch("django.conf.settings.UPLOAD_SESSION_EXPIRE_AFTER_INACTIVE_MINUTES", 60)
+    def test_saving_expirable_in_progress_submission(self) -> None:
```
Addressed by: https://github.com/NationalCentreTruthReconciliation/Secure-Record-Transfer/pull/442/commits/1b8249af08beaf24b7df572d558bbd95ef295ae2

danloveg reviewed (changes requested):
This looks good. The only thing is that the emails are trying to be called, and a connection to redis is failing to be established, causing some tests to fail.

Maybe you can patch/mock the email functions to skip trying to call them as jobs?

yenaing-oo commented:
> This looks good. The only thing is that the emails are trying to be called, and a connection to redis is failing to be established, causing some tests to fail.
> 
> Maybe you can patch/mock the email functions to skip trying to call them as jobs?

Strangely, when I ran `test_wizard` using the VSCode play button, it ran without error, but when I ran `pytest -k "unit"`, only then did the mentioned error show up.  I'll have to make sure to double-check with the command before pushing.

I've patched the email jobs like you suggested here: https://github.com/NationalCentreTruthReconciliation/Secure-Record-Transfer/pull/442/commits/5dc4b034c7bbac3517a44eac1cdd08e8f86c67ca

I also fixed a zero-padding mismatch in `test_profile` for the hour. The actual message does not use zero-padding for the hour but the test was expecting it to. Addressed by: https://github.com/NationalCentreTruthReconciliation/Secure-Record-Transfer/pull/442/commits/0e1f18d9ee5a21e8bfe0a6aae1d7be3f02c21f96

danloveg commented on app/recordtransfer/tests/unit/views/test_profile.py:
```diff
@@ -287,7 +287,7 @@ def test_in_progress_submission_expires_at(self) -> None:
         local_tz = ZoneInfo(settings.TIME_ZONE)
         expiry_date = self.upload_session.expires_at.astimezone(local_tz).strftime(
-            "%a %b %-d, %Y @ %-H:%M"
+            "%a %b %-d, %Y @ %H:%M"
```
I'm using Python 3.12.6 locally and get this error when running this test:

```
self = <recordtransfer.tests.unit.views.test_profile.TestUserProfileView testMethod=test_in_progress_submission_expires_at>

    @patch("django.conf.settings.UPLOAD_SESSION_EXPIRE_AFTER_INACTIVE_MINUTES", 60)
    @patch("django.conf.settings.UPLOAD_SESSION_EXPIRING_REMINDER_MINUTES", 30)
    def test_in_progress_submission_expires_at(self) -> None:
        """Test that expiry date is shown for an in-progress submission with an upload session."""
        response = self.client.get(self.url)
        local_tz = ZoneInfo(settings.TIME_ZONE)
>       expiry_date = self.upload_session.expires_at.astimezone(local_tz).strftime(
            "%a %b %-d, %Y @ %H:%M"
        )
E       ValueError: Invalid format string
```

The fix I made in https://github.com/NationalCentreTruthReconciliation/Secure-Record-Transfer/pull/444/commits/f4b14ff558f258b25f9605fa5921ff0942059bcc is a more robust way to handle the leading zeroes. It's a bit harder to read is the only thing.

yenaing-oo replied:
I see, I've merged your fix from `master` into this branch here: https://github.com/NationalCentreTruthReconciliation/Secure-Record-Transfer/pull/442/commits/a0243064a73b1224e4a4c6cf823837caabad9b32

danloveg commented on app/recordtransfer/tests/unit/views/test_transfer.py:
```diff
@@ -15,32 +17,221 @@ def setUp(self) -> None:
+                response = self.client.post(self.url, submit_data, follow=True)
+                self.assertEqual(200, response.status_code)
+                self.assertRedirects(response, reverse("recordtransfer:userprofile"))
+                self.assertContains(
+                    response, "Transfer saved successfully. This submission will expire on"
+                )
```
This is failing for me when running the tests locally for some reason. But it's not failing when it runs in the container. Does this pass for you when you run the tests locally?

yenaing-oo replied:
Yes, all of the tests pass locally for me.

danloveg replied:
Weird, must just be me then.
